modules/product/src/Form/AddToCartForm.php

```diff
@@ -166,4 +182,253 @@ function submitForm(array &$form, FormStateInterface $form_state) {
     ]));
   }
 
+  /**
+   * Builds the variation selection form.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   * @param \Drupal\commerce_product\Entity\ProductVariationInterface[] $variations
+   *   An array of product variations.
+   */
+  protected function variationForm(array &$form, FormStateInterface $form_state, array $variations) {
+    // Resolve the default variation from current attribute values.
+    $this->resolveDefaultVariation($variations, $form_state);
+
+    // Get qualified attribute options and add them to the form.
+    foreach ($this->resolveQualifyingAttributes($variations) as $name => $data) {
+      $this->attributeElement($form, $data, $data['used_options']);
+    }
+
+    if (!empty($form['attributes'])) {
+      $form['attributes'] += [
+        '#tree' => TRUE,
+        '#prefix' => '<div id="attribute-widgets" class="attribute-widgets">',
+        '#suffix' => '</div>',
+        '#weight' => 0,
+      ];
+      $form['unchanged_attributes'] += [
+        '#tree' => 'TRUE',
+      ];
+    }
+  }
+
+  /**
+   * Adds an attribute form element to add to cart form.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param array $attribute
+   *   The attribute information array.
+   * @param $used_options
+   *   An array of options in use.
+   */
+  protected function attributeElement(array &$form, array $attribute, array $used_options) {
+    $field_name = $attribute['name'];
+    $options = array_intersect_key($attribute['options'], array_combine($used_options[$field_name], $used_options[$field_name]));
```

Comment:
The parent method passes $attribute['options'] as $used_options, so this array_intersect_key does nothing.